Add purview workflow (#33629)
Azure/azure-sdk-for-net: sdk/purview/Azure.Analytics.Purview.Workflows/README.md

# Azure Purview Workflow client library for .NET

Workflows are automated, repeatable business processes that users can create within Microsoft Purview to validate and orchestrate CUD (create, update, delete) operations on their data entities. Enabling these processes allow organizations to track changes, enforce policy compliance, and ensure quality data across their data landscape.

Use the client library for Purview Workflow to:
- CRUD of workflow.
- Submit user request and monitor workflow run.
- View and respond to workflow tasks.

**Please rely heavily on the [service's documentation][product_documentation]**

## Getting started

> This section should include everything a developer needs to do to install and create their first client connection *very quickly*.

Author: This section was reserved for the sdk download address, I have creates the issue for tracking completing README, here is issue link: https://github.com/Azure/azure-sdk-for-net/issues/33756

### Authenticate the client

Since Workflow service use Azure Active Directory bearer token for authentication and username(usually it is an email address) should be encoded into the token so that user could received email notification when using workflow. A recommended solution is using the [Azure Identity][azure_identity] library to provide the [UsernamePasswordCredential][username_password_credentail]. Before get started using the [Azure Identity][azure_identity] library, [an application][app_registration] should be registered and then provide the client id into the [UsernamePasswordCredential][username_password_credentail].

```C#
    //<client-id>, <tenant-id> are the client ID and tenant ID of the AAD application.
    //<user-name>,<password> are the value of user name and password of an AAD user.
    TokenCredential usernamePasswordCredential = new UsernamePasswordCredential(<client-id>, <tenant-id>, <user-name>,<password>, null);
    var client = new PurviewWorkflowServiceClient(new Uri("https://<purview-account-name>.purview.azure.com"), usernamePasswordCredential)
```

## Key concepts

> The *Key concepts* section should describe the functionality of the main classes. Point out the most important and useful classes in the package (with links to their reference pages) and explain how those classes work together. Feel free to use bulleted lists, tables, code blocks, or even diagrams for clarity.

Author: Already tracked in issue, issue link: https://github.com/Azure/azure-sdk-for-net/issues/33756

### Additional concepts
<!-- CLIENT COMMON BAR -->
[Client options](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/README.md#configuring-service-clients-using-clientoptions) |
[Accessing the response](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/README.md#accessing-http-response-details-using-responset) |
[Long-running operations](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/README.md#consuming-long-running-operations-using-operationt) |
[Handling failures](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/README.md#reporting-errors-requestfailedexception) |
[Diagnostics](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/Diagnostics.md) |
[Mocking](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/README.md#mocking) |
[Client lifetime](https://devblogs.microsoft.com/azure-sdk/lifetime-management-and-thread-safety-guarantees-of-azure-sdk-net-clients/)
<!-- CLIENT COMMON BAR -->

## Examples

> You can familiarize yourself with different APIs using [Samples](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/purview/Azure.Analytics.Purview.Workflows/samples).

Author: Thanks for your feedback, I've already added examples in README.

### <scenario>

You can create a client and call the client's `<operation>` method.

```C# Snippet:Azure_Analytics_Purview_Workflows_CrudWorkflow
Uri endpoint = new("https://<purview-account-name>.purview.azure.com");
var credential = new DefaultAzureCredential();
var client = new PurviewWorkflowServiceClient(endpoint, credential);

Guid workflowId = new Guid("ba25ed0e-3364-4e8e-8385-c60e12f3e342");
Response result = client.GetWorkflow(workflowId);
```

## Troubleshooting

> Describe common errors and exceptions, how to "unpack" them if necessary, and include guidance for graceful handling and recovery.

Author: Already tracked in issue, issue link: https://github.com/Azure/azure-sdk-for-net/issues/33756

Use the client library for Purview Workflow to:
- Managing workflows
- Submitting user requests and monitoring workflow runs
- Viewing and responding to workflow tasks

**For more details about how to use workflow, please refer to the [service documentation][product_documentation]**

### Prerequisites

You must have an [Azure subscription][azure_subscription] and a [Purview resource][purview_resource] to use this package.

### Authenticate the client

Since the Workflow service uses an Azure Active Directory (AAD) bearer token for authentication and identification, an email address should be encoded into the token to allow for notification when using Workflow. It is recommended that the [Azure Identity][azure_identity] library be used  with a the [UsernamePasswordCredential][username_password_credential]. Before using the [Azure Identity][azure_identity] library with Workflow, [an application][app_registration] should be registered and used for the clientId passed to the [UsernamePasswordCredential][username_password_credentail].

> ```C# Snippet:Azure_Analytics_Purview_Workflows_CreatePurviewWorkflowClient

Reviewer: Run `eng\scripts\Update-Snippets.ps1` to fill the real content into this snippet.

Author: Done

# Azure Purview Workflow client library for .NET

Workflows are automated, repeatable business processes that users can create within Microsoft Purview to validate and orchestrate CUD (create, update, delete) operations on their data entities. Enabling these processes allow organizations to track changes, enforce policy compliance, and ensure quality data across their data landscape.

Use the client library for Purview Workflow to:

> - Managing workflows
> - Submitting user requests and monitoring workflow runs
> - Viewing and responding to workflow tasks
> 

Reviewer 2: ```suggestion
- Manage workflows
- Submit user requests and monitoring workflow runs
- View and respond to workflow tasks

```

Response createResult = await client.CreateOrReplaceWorkflowAsync(workflowId, RequestContent.Create(workflow));
```

### Submit user requests

```C# Snippet:Azure_Analytics_Purview_Workflows_SubmitUserRequests
string request = "{\"operations\":[{\"type\":\"CreateTerm\",\"payload\":{\"glossaryTerm\":{\"name\":\"term\",\"anchor\":{\"glossaryGuid\":\"20031e20-b4df-4a66-a61d-1b0716f3fa48\"},\"status\":\"Approved\",\"nickName\":\"term\"}}}],\"comment\":\"Thanks!\"}";

Response submitResult = await client.SubmitUserRequestsAsync(RequestContent.Create(request));
```

### Approve workflow task

```C# Snippet:Azure_Analytics_Purview_Workflows_ApproveWorkflowTask

> // This taskId is an existing workflow task's id, user could get workflow tasks by calling GetWorkflowTasksAsync API.

Reviewer 2: ```suggestion
// This taskId represents an existing workflow.  The id can be obtained by calling GetWorkflowTasksAsync.
```

Guid taskId = new Guid("b129fe16-72d3-4994-9135-b997b9be46e0");

string request = "{\"comment\":\"Thanks!\"}";

Response approveResult = await client.ApproveApprovalTaskAsync(taskId, RequestContent.Create(request));
```

<!-- LINKS -->
[product_documentation]: https://learn.microsoft.com/azure/purview/concept-workflow
[azure_subscription]: https://azure.microsoft.com/free/dotnet/
[purview_resource]: https://docs.microsoft.com/azure/purview/create-catalog-portal

> [azure_identity]: https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity/Azure.Identity

Reviewer 2: ```suggestion
[azure_identity]: https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity/Azure.Identity/README.md
```
